contrib/hamilton/contrib/user/skrawcz/fine_tuning/__init__.py

```diff
@@ -0,0 +1,653 @@
+import json
+import logging
+from functools import partial
+from typing import Dict
+
+logger = logging.getLogger(__name__)
+
+from hamilton import contrib
+
+with contrib.catch_import_errors(__name__, __file__, logger):
+    import evaluate
+    from datasets.combine import DatasetType
+    from datasets import Dataset, concatenate_datasets
+    import numpy as np
+    import pandas as pd
+    from peft import (
+        LoraConfig,
+        PeftConfig,
+        PeftModel,
+        TaskType,
+        get_peft_model,
+        prepare_model_for_int8_training,
+    )
+    from sklearn.model_selection import train_test_split
+    import torch
+    from tqdm import tqdm
+    from transformers import (
+        AutoModelForSeq2SeqLM,
+        AutoTokenizer,
+        DataCollatorForSeq2Seq,
+        Seq2SeqTrainer,
+        Seq2SeqTrainingArguments,
+        PreTrainedTokenizerBase,
+    )
+
+from hamilton.function_modifiers import extract_fields
+from hamilton.function_modifiers.configuration import config
+
+
+@extract_fields({"train_set": Dataset, "validation_set": Dataset, "inference_set": Dataset})
+def raw_dataset(
+    data_path: str,
+    random_state: int = 42,
+    train_size: float = 0.95,
+    validation_size: float = 0.8,
+    input_text_key: str = "question",
+    output_text_key: str = "reply",
+) -> Dict[str, Dataset]:
+    """Loads the raw dataset from disk and splits it into train and test sets.
+
+    :param data_path: the path to the dataset.
+    :param random_state: the random state for the train/test split.
+    :param train_size: the train size as a fraction of the dataset.
+    :param input_text_key: the key to use to get the input text from what was loaded.
+    :param output_text_key: the key to use to get the output text from what was loaded.
+    :return: dictionary with the train, validation, and inference sets.
+    """
+    with open(data_path, "r") as f:
+        data = json.load(f)
+
+    format_data = []
+    for sample in data:
+        format_data.append(
+            {
+                "input_text": sample[input_text_key],
+                "output_text": sample[output_text_key],
+            }
+        )
+    format_data = pd.DataFrame(format_data)
+    train, test = train_test_split(format_data, random_state=random_state, train_size=train_size)
+    # split the test set into a validation set and inference set
+    validation, inference = train_test_split(
+        test, random_state=random_state, train_size=validation_size
+    )
+    dataset_train = Dataset.from_pandas(train)
+    dataset_validation = Dataset.from_pandas(validation)
+    dataset_inference = Dataset.from_pandas(inference)
```

Comment:
Why an inference dataset? Feels to me that that's a different pipeline -- E.G. an inference pipeline. You'll want to train the model then infer on a dataset. That'll make more sense and make it reusable. 

If we remove this we get more training/test data, and we can just add a quick repl to load the model we saved to make an inference request.

Reply:
because it's a hold out set. If you're going to eval performance it should be on an unseen set.

Reply:
can rename to hold out set.

Reply:
Yes, this relates to my whole point about too many things in this pipeline. But yeah, a hold-out set is fine.

Reply:
Remember Hamilton allows you to choose what you compute so extra stuff isn't an issue.

In terms of code organization we should consider extending the hub to allow multiple modules in one download, since I agree this could be split up into multiple modules, but in terms of a full example I think this is the best way for now.